virtual_ecosystem/models/soil/microbial_groups.py

```diff
@@ -0,0 +1,138 @@
+"""The ``models.soil.microbial_groups`` module contains the classes needed to define
+the different microbial functional groups used in the soil model.
+"""  # noqa: D205
+
+from dataclasses import dataclass
+
+from virtual_ecosystem.models.soil.constants import SoilConsts
+
+
+@dataclass
+class MicrobialGroup:
```

Comment:
If I get this right, this dataclass is mean to contain the constants that each group (bacteria and fungi) requires, and is not meant to change over the course of the simulation, right?

If that's the case, I'd make this a frozen dataclass, to ensure that, and maybe add `Constants` to the class name to make it even more clear:
```suggestion
@dataclass(frozen=True)
class MicrobialGroupConstants:
```

Reply:
Yep these constants should never change in the course of a simulation, I'll make it a frozen class and rename!

Reply:
At the moment you already have all of this data packaged in `SoilConsts` but you are packaging it up into a consistent wrapper and leaving the door open to add more groups? I think as it stands its between two stools - your code allows multiple groups but the config has got you pinned at only one of each.

I think you've basically got microbial functional types here and it would make sense to adopt something like the framework used in the Plants Model update. Rather than having a long list of of parameters which (currently) only allow one of each of fungi and bacteria, then have an array of tables in TOML (or package in a CSV file as Plants does).

```toml
[[soil.microbial_group]]
type = "fungal"
max_uptake_rate_labile_C = 0.1
etc = "etc"
```

If you add `type` to your dataclass, you can then simply load an instance using `MicrobialGroup(**args_from_config)` and hence ditch `make_bacterial_functional_group` and `make_fungal_functional_group`. The only thing you then need to do is to divide what you get from loading the config into the fungal and bacterial subgroups in `make_full_set_of_microbial_groups`.

Reply:
That does make sense, I guess my main concern would be how these constants will get documented because the majority of "constants" used in the soil model in reality vary between functional groups.

Are they documented as attributes in `MicrobialGroup` rather than in the constants submodule?

Reply:
So you have a consistent set of attributes across both fungi and bacteria but the values of those attributes differ between the two (and also _could_ vary between different subgroups within those two, but currently don't).

So, I think the attribute itself should be easy to document - it plays a consistent role - but then the actual values for each group are harder. Ultimately this is becoming more like data - as a user, you'd want to document what sources were used to describe those groups but that isn't something that we document within the code? 

I guess you'd also need to handle the case where users omit one of the two! Could you run the simulation without one of them or should it break?

Reply:
It should break if one of the two is omitted. It's not implemented yet but the idea is that each functional groups interacts with the soil pools (somewhat) differently. With the basic idea being the functional groups are defined at a coarse enough level that they represent really meaningfully different actors, as the model is not really designed to simulate competition between similar organisms.

I really don't want to allow users to define arbitrary functional groups, because this would essentially entail spending a lot of development effort for a feature we would tell people not to use (because the results would be nonsensical, like if two bacterial functional groups were added one would be guaranteed to go extinct).

Reply:
I do think it's worth adopting your approach though as it will save me time and effort (and avoid cluttering constants) when I get around to split fungi into multiple groups

Reply:
> I really don't want to allow users to define arbitrary functional groups

Is that then a simple check that the `types` are unique on the loaded set of groups?

> spending a lot of development effort for a feature we would tell people not to use 

But we do want them to use it - it's a less cluttered way of defining the constants. We just don't want them to use it wrong 😄 It also opens up using a CSV for the parameterisation, which is much more user friendly than TOML for this kind of configuration.